feat: Proxy Mask response transform (#435)
Basis-Theory/developers.basistheory.com: docs/api/proxies/pre-configured-proxies.mdx

### Response Transforms

> There are two types of Response Transforms available; `code` and `mask`.

Reviewer: ```suggestion
There are two types of Response Transforms available: `code` and `mask`.
```

};
```

> ##### Code Response Transform - Javascript Request Object

Reviewer: Just double-checking: did you review existing links in docs for these heading changes?

Author: I did a global search for "response-transform-request-object" and "response-transform-response-object" and found no results.

| `body`    | _object_ | The transformed response body to return from the proxy    |
| `headers` | _object_ | The transformed response headers to return from the proxy |

#### Mask Response Transform

> Mask response transforms can only be used to transform the body of the response received from the destination before returning a response from the proxy.

Reviewer: ```suggestion
Mask response transforms can be used to transform the body of the response received from the destination before returning a response from the proxy.
```

The mask transform has two available matcher types, `regex` and `chase_stratus_pan`.

##### Regex Matcher

Mask response transforms using the `regex` matcher are defined as regular expressions and a replacement character.

```javascript
{
    "name": "My masked proxy",
    "destination_url": "https://echo.basistheory.com/anything",
    "require_auth": false,
    "response_transform": {
        "type":"mask",
        "matcher":"regex",
        "replacement":"*",
        "expression":"\"method\":\\s*\"(.*?)\""
    }
}
```

The regular expression must contain at least one capture group.

> In the example above, the value of the `method` json attribute will be replaced with `*`.

Reviewer: This example confused me by relying on the Echo server behavior. My first thought was believing this was some sort of built-in mechanism to identify HTTP request properties (e.g. `method`). 
It may be worth it to add a sample response to work from. That way we can be more explicit about how this new feature works.

Author: Agree. I'll will try adding a more clearer example.

If the method were `GET`, the replacement would be `***`. 
If the method were `POST`, then `****` would be the replacement.
When a response body matches the given expression, the matched part of the body will be processed.
During processing, any portion of the matched body section that is equivalent to any value of a capture group defined in the regex will be replaced.

The best practice is to limit the match scope to be as narrow as possible.
There could be unintended or unexpected side effects if the match string is larger than necessary.

For example, given a regex defined as `^(aa).*?$`, a match will occur when the line starts with "aa".
However, since the regular expression uses the beginning, `^`, and end `$` syntax, the whole string will be matched and eligible for replacements.
Given an input string of `aabbccaabbccaa`, the entire string will processed during masking. 
*ALL* values that match the capture group value, in this case `aa`, will be replaced.
The string will then be masked as `**bbcc**bbcc**`.

##### Chase Stratus PAN Matcher

> If your proxy is interfacing with the Chase Stratus service, you may be able to take advantage of a pre-configured mask.

Reviewer: ```suggestion
If your proxy is interfacing with the Chase Stratus service, you may be able to take advantage of a default mask.
```

Avoid confusion with "pre-configured" concept applied to Proxies, which requires user customization.

        "type":"mask",
        "matcher":"regex",
        "replacement":"*",
        "expression":"\"accountNumber\":\\s*\"(.*?)\""
    }
}
```

The regular expression must contain at least one capture group.
In the example above, the `expression` will capture the value of an `accountNumber` attribute of a JSON payload.
Given the proxied service responds with a JSON similar to the following:

> ```json

Reviewer: ```suggestion
```json title="Original Response"
```

{
    "username": "bsmith1486",
    "accountNumber": "56834512" 
}
```
Then the sample proxy will mask the response as:

> ```json

Reviewer: ```suggestion
```json title="Transformed Response"
```
